subprojects/java-compiler-plugin/src/main/java/org/gradle/internal/compiler/java/IncrementalCompileTask.java

```diff
@@ -68,7 +68,7 @@ public void setLocale(Locale locale) {
     @Override
     public Boolean call() {
         if (delegate instanceof JavacTask) {
-            ClassNameCollector collector = new ClassNameCollector(relativize);
+            ClassNameCollector collector = new ClassNameCollector(relativize, ((JavacTask) delegate).getElements());
```

Comment:
Calling `javacTask.getElements()` here has a rather weird side effect on deprecation messages in Java 8 - see this commit: https://github.com/gradle/gradle/pull/16066/commits/b04b5aba841423afc70500446f0f8dbf62a5c88c
Not sure if there can be more implications here.

Reply:
I think the side effect is not too bad but would be good to capture in the tests instead of ignoring it. Initially, I thought they dropped the class name completely which is not great.

Reply:
And would be nice to pull the cast out as we do the same cast in the next line. Time to switch to Java 14 with pattern matching :)

Reply:
I wonder also if the `instanceof-or-throw` can be pulled out to the constructor so that the `delegate` field is always a `JavacTask` here. Let's see what CI thinks.